python/cudf/cudf/utils/_setup_numba.py

```diff
@@ -0,0 +1,160 @@
+        )
+    else:
+        return regular_result[1]
+
+
+def _setup_numba_linker(path):
+    """
+    Configure the numba linker for use with cuDF. This consists of
+    potentially putting numba into enhanced compatibility mode
+    based on the user driver and runtime versions as well as the
+    version of the CUDA Toolkit used to build the PTX files shipped
+    with the user cuDF package.
+    """
+    try:
+        # By default, ptxcompiler will not be installed with CUDA 12
+        # packages. This is ok, because in this situation putting
+        # numba in enhanced compatibility mode is not necessary.
+        from cubinlinker.patch import _numba_version_ok, new_patched_linker
+        from ptxcompiler.patch import NO_DRIVER, safe_get_versions
+    except ImportError:
+        return
+    versions = safe_get_versions()
+    if versions != NO_DRIVER:
+        driver_version, runtime_version = versions
+        # Don't check if CEC is necessary in the possible edge
+        # case where a user has a CUDA 12 package and ptxcompiler
+        # in their environment anyways, perhaps installed separately
+        if driver_version < (12, 0):
+            ptx_toolkit_version = _get_cuda_version_from_ptx_file(path)
+            # Numba thinks cubinlinker is only needed if the driver is older
+            # than the CUDA runtime, but when PTX files are present, it might
+            # also need to patch because those PTX files may be compiled by
+            # a CUDA version that is newer than the driver as well
+            if (driver_version < ptx_toolkit_version) or (
+                driver_version < runtime_version
+            ):
+                if _numba_version_ok:
+                    Linker.new = new_patched_linker
+
+
+def _get_cuda_version_from_ptx_file(path):
+    """
+    https://docs.nvidia.com/cuda/parallel-thread-execution/
+    Each PTX module must begin with a .version
+    directive specifying the PTX language version
+
+    example header:
+    //
+    // Generated by NVIDIA NVVM Compiler
+    //
+    // Compiler Build ID: CL-31057947
+    // Cuda compilation tools, release 11.6, V11.6.124
+    // Based on NVVM 7.0.1
+    //
+
+    .version 7.6
+    .target sm_52
+    .address_size 64
+
+    """
+    with open(path) as ptx_file:
+        for line in ptx_file:
+            if line.startswith(".version"):
+                ver_line = line
+                break
+        else:
+            raise ValueError("Could not read CUDA version from ptx file.")
+    version = ver_line.strip("\n").split(" ")[1]
+    # from ptx_docs/release_notes above:
+    ver_map = {
+        "7.5": (11, 5),
+        "7.6": (11, 6),
+        "7.7": (11, 7),
+        "7.8": (11, 8),
+        "8.0": (12, 0),
+    }
+
+    cuda_ver = ver_map.get(version)
+    if cuda_ver is None:
+        raise ValueError(
+            f"Could not map PTX version {version} to a CUDA version"
+        )
+
+    return cuda_ver
+
+
+class CUDFNumbaConfig:
```

Comment:
I don't like that this class is in a file named `_setup_numba`. It's not something used at import time, but at runtime when calling Numba. Should this be in a "utils"? Maybe rename this file to `utils/_numba.py` since it contains both import logic and runtime logic? Otherwise I would like the scope of `_setup_numba.py` to _only_ be import-time logic as implied by the name.

Also make it private.
```suggestion
class _CUDFNumbaConfig:
```